AirCasting/SessionViews/ChartViewModel.swift

```diff
@@ -0,0 +1,96 @@
+// Created by Lunar on 30/09/2021.
+//
+
+import Foundation
+import Charts
+
+final class ChartViewModel: ObservableObject {
+    @Published var entries: [ChartDataEntry] = []
+    
+    private let stream: MeasurementStreamEntity
+
+    private var timeUnit: Double {
+        stream.session.type == .mobile ? 60 : 60 * 60
+    }
+    
+    private var mainTimer: Timer?
+    private var firstTimer: Timer?
+    
+    deinit {
+        mainTimer?.invalidate()
+        firstTimer?.invalidate()
+    }
+    
+    init(stream: MeasurementStreamEntity) {
+        self.stream = stream
+        generateEntries()
+        startTimers(stream.session)
+    }
+    
+    private func startTimers(_ session: SessionEntity) {
+        let timeOfNextAverage = timeOfNextAverage()
+
+        firstTimer = Timer.scheduledTimer(withTimeInterval: TimeInterval(timeOfNextAverage), repeats: false) { [weak self] timer in
+            self?.generateEntries()
+            self?.startMainTimer()
+        }
+    }
+    
+    func startMainTimer() {
+        mainTimer = Timer.scheduledTimer(withTimeInterval: timeUnit, repeats: true) { [weak self] timer in
+            self?.generateEntries()
+        }
+    }
+    
+    private func generateEntries() {
+        // Set up begning and end of the interval for the first average
+        //  - for fixed sessions we are taking the last full hour of the session, which has any measurements
+        //  - for mobile we are taking into account full minutes since the session started and we are taking the most recent one
+        guard var intervalEnd = endTime() else {
+            return
+        }
+        var intervalStart = intervalEnd - timeUnit
+        
+        var entries = [ChartDataEntry]()
+        
+        for i in (0...8).reversed() {
+            if (intervalStart < stream.session.startTime!) { break }
+            let average = averagedValue(intervalStart, intervalEnd)
+            if let average = average {
+                entries.append(ChartDataEntry(x: Double(i), y: average))
+            }
+            intervalEnd = intervalStart
+            intervalStart = intervalEnd - timeUnit
+        }
+        
+        self.entries = entries
+    }
+    
+    private func endTime() -> Date? {
```

Comment:
I'd name it intervalEndTime() - endTime name may be confused with session endTime